### PR TITLE
perf: better L2 code generation

### PR DESCRIPTION
**Pull request by eddyxu**
* 7-10% faster in L2 benchmark
* 6% faster in hnsw build.

### CODE REVIEW

**Comment by eddyxu** [comment]
```
L2(f32, auto-vectorization)
                        time:   [90.263 ms 90.800 ms 91.843 ms]
                        change: [-9.2168% -8.4476% -7.1232%] (p = 0.00 < 0.10)
                        Performance has improved.
                        
create_hnsw(10240x512,levels=6)
                        time:   [718.37 ms 744.59 ms 778.73 ms]
                        change: [-10.879% -6.8104% -1.6829%] (p = 0.02 < 0.05)
                        Performance has improved.                      
```

File: rust/lance-linalg/src/distance/l2.rs [hunk 1]
```diff
@@ -150,7 +150,7 @@ impl L2 for f16 {
 impl L2 for f32 {
     #[inline]
     fn l2(x: &[Self], y: &[Self]) -> f32 {
-        l2_scalar::<Self, Self, 32>(x, y)
+        l2_scalar::<Self, Self, 16>(x, y)
```
Comment (wjones127) [inline rust/lance-linalg/src/distance/l2.rs:153]: How did you choose this value? Could this be overfitting for the CPU you are benchmarking on?

Comment (eddyxu) [reply rust/lance-linalg/src/distance/l2.rs:153]: I tested in both on m2 and an amd 5950x machine.  Also tried to generate code with https://rust.godbolt.org/z/xoWf47WqE, with `--target-cpu=cascadelake`.

so `16 = 512 / 8 bit / 4 (f32)`.

Comment (wjones127) [reply rust/lance-linalg/src/distance/l2.rs:153]: Could you leave a comment there so we remember this?

**Review by wjones127 — APPROVED** [review]

File: rust/lance-linalg/src/distance/l2.rs (hunk 1 above)
Comment (eddyxu) [reply rust/lance-linalg/src/distance/l2.rs:155]: Sure. done.

**Pull request merged**